### PR TITLE
Temperature output for producers

### PR DESCRIPTION
**Pull request by plgbrts**
This is PR for temperature output reporting for producers. This is especially relevant for geothermal reservoir simulations.

A comparison with eclipse is given below for a three phase system:
image

Applied expressions derived for average well temperature: 
image

### CODE REVIEW

**Comment by totto82** [comment]
Thanks for the contribution. As I understand the code this compute a well temperature based on a weighted average of the temperature in the perforated cells. 

The output temperature is computed when initializing the well i.e. in the beginning of a report step. Don't you want the temperature at the end of each timestep?

**Comment by plgbrts** [comment]
> Thanks for the contribution. As I understand the code this compute a well temperature based on a weighted average of the temperature in the perforated cells.
> 
> The output temperature is computed when initializing the well i.e. in the beginning of a report step. Don't you want the temperature at the end of each timestep?


Indeed the  reported temperature is a weighted average of the temperature in the perforated cells.

I have noticed that the temperature reporting is  limited to the reporting timesteps but I am not sure where to put the temperature reporting  in such way it is reported at each sub timestep as you suggest. Currently I get this  output profile (comparison with eclipse, see picture below) matching temperatures at reporting timesteps but not at the sub timesteps.
If you can advise me on how to report temperatures at each timestep it would be very much appreciated.


image

**Comment by totto82** [comment]
I suggest to add a computeWellTemperature(...)  to BlackoilWellModel_impl.hpp 
that is called from timeStepSucceeded(...) in BlackoilWellModel_impl.hpp
and move your calculations from the wellState.init(...) function to computeWellTemperature

Pseudo code for computeWellTemperature(). 
I suggest that you quarry the fluidState for each perforation and thus avoid looping over the whole grid. 
as shown here:
```
if !has_energy
  return

for well in wells()
  wellTemp = 0
  if (well.injector())
     continue;
  for perf in perfs()
        // get the intensive quantifies just for the cell_idx of the perforations
        const auto& intQuants = *(ebosSimulator.model().cachedIntensiveQuantities(cell_idx, /*timeIdx=*/ 0));
        const auto& fs = intQuants.fluidState();

        // do your stuff
        ...........
  wellState.temperature() = wellTemp

```

File: opm/simulators/wells/WellState.hpp [hunk 1]
```diff
@@ -373,6 +383,32 @@ namespace Opm
...
+                   for (int i = 0; i < num_perf_this_well; ++i){
+                        int perf_cell_index = well_perf_data_[w][i].cell_index;
+                        double weight_factor = 0.0;
+                        for (int p = 0; p < np; ++p) {     
+                            weight_factor += cellDensity[perf_cell_index][p]  * perforationRates[connpos + i*np + p]/cellBinv[perf_cell_index][p] * cellInternalEnergy[perf_cell_index][p]/cellTemperatures[perf_cell_index];
+                        }
+                        total_weight += weight_factor;
+                        weighted_temperature += weight_factor * cellTemperatures[perf_cell_index];
+                    }
+                    temperature_[w] = weighted_temperature/total_weight;
```
Comment (blattms) [inline opm/simulators/wells/WellState.hpp:405]: We need a 
```
weighted_temperature = well_info.communication().sum(weighted_temperature)
total_weight = well_info.communication().sum(total_weight);
```
before the division to account for wells which have perforation on different processes.

**Comment by plgbrts** [comment]
> I suggest to add a computeWellTemperature(...) to BlackoilWellModel_impl.hpp
> that is called from timeStepSucceeded(...) in BlackoilWellModel_impl.hpp
> and move your calculations from the wellState.init(...) function to computeWellTemperature
> 
> Pseudo code for computeWellTemperature().
> I suggest that you quarry the fluidState for each perforation and thus avoid looping over the whole grid.
> as shown here:
> 
> ```
> if !has_energy
>   return
> 
> for well in wells()
>   wellTemp = 0
>   if (well.injector())
>      continue;
>   for perf in perfs()
>         // get the intensive quantifies just for the cell_idx of the perforations
>         const auto& intQuants = *(ebosSimulator.model().cachedIntensiveQuantities(cell_idx, /*timeIdx=*/ 0));
>         const auto& fs = intQuants.fluidState();
> 
>         // do your stuff
>         ...........
>   wellState.temperature() = wellTemp
> ```

Many thanks for this good and very helpful  suggestion it makes it a much cleaner implementation and solves the reporting issue for the sub timesteps.

File: opm/simulators/wells/WellState.hpp (hunk 1 above)
Comment (plgbrts) [reply opm/simulators/wells/WellState.hpp]: Thanks for pointing out this. I have included your suggestion.

Comment (blattms) [reply opm/simulators/wells/WellState.hpp]: Thanks a lot.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
Looks good. Lets get this merged.

**Comment by totto82** [comment]
Thanks for the contribution.

**Pull request merged**